src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs

```diff
@@ -125,6 +125,8 @@ public StringSegment(string buffer, int offset, int length)
         /// <exception cref="ArgumentOutOfRangeException">
         /// <paramref name="start"/> is greater than or equal to <see cref="Length"/> or less than zero.
         /// </exception>
+        [MemberNotNull(nameof(Buffer))]
+        [MemberNotNull(nameof(Value))]
```

Comment:
The method doesn't make either property non-`null`; it just `throw`s when either are `null` and there's no `[DoesNotReturnIfMemberNull(...)]` attribute available. Suggest leaving the metadata here alone.

Same for the other `AsSpan(...)` overloads. (Side note: I doubt the compiler will lift metadata to callers.)

----

For example, compiler will think the following is just fine.
``` c#
    var segment = StringSegment.Empty;
    try
    {
        var span = segment.AsSpan(23);
        // use span
    }
    catch
    {
    }

    string value = segment.Value;
    // or
    var length = segment.Value.Length;
```

Reply:
`MemberNotNull` just means that member won't be null when method returns (throwing doesn't count). Since this method throws when `Value` is null, if it returns - it is not null.

Example you provided seems to be incorrect:

![image](https://user-images.githubusercontent.com/6609929/131228084-b401dc2d-19ac-4dd1-bda4-2b8a9e944db5.png)

From compiler's point of view if method throws it didn't finish executing so there's no guarantee that member is initialized.

Reply:
From the [doc](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.membernotnullattribute?view=net-5.0):
> Specifies that the method or property will ensure that the listed field and property members have values that aren't null.

Method doesn't need to modify members, just ensure that they are initialized.